File: Javascript/roman-to-int/src/RomanToInt.js

function RomanToInt(sampStr){
    this.digitVal = {
        'i' : 1,
        'v' : 5,
        'x' : 10,
        'l' : 50,
        'c' : 100,
        'd' : 500,
        'm' : 1000,
    };
    this.sampStr = sampStr;
    this.intVal = 0;
}

RomanToInt.prototype.print = function(){
    console.log(`The integer form of roman: '${this.sampStr}' is: ${this.intVal}`);
}



RomanToInt.prototype.convert = function (){
    let romanString = this.sampStr.toLowerCase();
    const romanMap = new Map();
    for (let i = 0; i < romanString.length; i++){
        romanMap.set( i, this.digitVal[romanString.charAt(i)] );
    }

    let adder = 0;
    while (romanMap.size !== 0){
        const romanIt = romanMap[Symbol.iterator]();
        let firstRoman = romanIt.next().value;
        let secondRoman = romanIt.next().value;

        if ( !(secondRoman && secondRoman[0] && secondRoman[1]) ){
            adder += firstRoman[1];
            break;
        }

        if( (firstRoman[1] > secondRoman[1]) || (firstRoman[1] == secondRoman[1]) ){
            adder += firstRoman[1];
            romanMap.delete(firstRoman[0]);
        }
        else{
            adder += secondRoman[1] - firstRoman[1];
            romanMap.delete(firstRoman[0]);
            romanMap.delete(secondRoman[0]);
        }
    }
    this.intVal = adder;
    return this;
};

module.exports = RomanToInt;

File: Java/RomanToInt/src/main/java/org/horbola/leetcode/java/RomanToInt.java
package org.horbola.leetcode.java;

import java.util.*;

public class RomanToInt {
    Map<Character, Integer> digitVal;
    String sampString = "";
    int intVal = 0;

    RomanToInt(String sampString){
        this.sampString = sampString;
        digitVal = new TreeMap();
        digitVal.put('i', 1);
        digitVal.put('v', 5);
        digitVal.put('x', 10);
        digitVal.put('l', 50);
        digitVal.put('c', 100);
        digitVal.put('d', 500);
        digitVal.put('m', 1000);
    }

    public static void main(String[] args) {
        for (;;){
            System.out.println("\n");
            System.out.println("Please enter the roman number to convert to integer");
            System.out.println("If you want to close the program, please press 'c'");
            Scanner in = new Scanner(System.in);
            String input = in.nextLine();
            if (input.equalsIgnoreCase("c"))
                break;
            else {
                callProgram(input);
            }
        }
    }

    private static void callProgram(Object input){
        String str = (String) input;
        var romanInt = new RomanToInt(str);
        romanInt.convert().print();
    }

    public void print(){
        if (this.intVal == 0){
            System.out.println("Sorry! error occurred.");
        }
        else {
            String str = String.format("The Integer form of the %s is: %s", this.sampString, this.intVal);
            System.out.println(str);
        }
    }


    RomanToInt convert(){
        String romanString = this.sampString.toLowerCase();
        Map<Integer, Integer> romanMap = new TreeMap();

        for(int i = 0; i < romanString.length(); i++){
            // convert roman letter to integer separately and put the integer
            Integer intVal = digitVal.get(romanString.charAt(i));
            romanMap.put(i, intVal);
        }

        int adder = 0;
        Set<Map.Entry<Integer, Integer>> romanEntrySet = romanMap.entrySet();
        while(!romanEntrySet.isEmpty()){
            Iterator it = romanEntrySet.iterator();
            Map.Entry<Integer, Integer> firstRoman = (Map.Entry<Integer, Integer>)it.next();
            Map.Entry<Integer, Integer> secondRoman = (Map.Entry<Integer, Integer>)it.next();

            if(!it.hasNext()){
                adder += firstRoman.getValue();
                break;
            }

            if( (firstRoman.getValue() > secondRoman.getValue()) || (firstRoman.getValue() == secondRoman.getValue()) ){
                adder += firstRoman.getValue();
                romanMap.remove(firstRoman.getKey());
            }
            else{
                adder += secondRoman.getValue() - firstRoman.getValue();
                romanMap.remove(firstRoman.getKey());
                romanMap.remove(secondRoman.getKey());
            }
        }
        this.intVal = adder;
        return this;
    }

}
